### PR TITLE
Add ability to find tests with different prefixes

### PR DESCRIPTION
**Pull request by schalkwijk**
Hello!

Some testing frameworks for Swift / Objective-C use different test prefixes to identify tests. For instance, [Quick](https://github.com/Quick/Quick) uses the `spec` method name prefix instead of `test`. In order to be able to use the `test-center` fastlane plugin with `Quick`, we need to be able to identify these differently-named tests, and so this commit introduces the ability to override the prefix that `XCTestList` uses to find tests.

I'm not entirely sure I tested this at the right level, so any feedback welcome. Note that I utilized the `match` method on the regex because, for some reason, interpolating the `test_prefix` directly in to the existing regex ended up breaking the capture group functionality, so this was my way around it. Feedback on that is, of course, welcome as well.

### CODE REVIEW

**Comment by lyndsey-ferguson** [comment]
@schalkwijk thank you for the PR, this is very exciting! I won't be able to look at this right away: hopefully sometime this week. If you're looking for a way to integrate this into test_center, you may want to look at the [`TestCollector `](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/lib/fastlane/plugin/test_center/helper/test_collector.rb) class.

**Comment by schalkwijk** [comment]
@lyndsey-ferguson, excellent! I'll try my hand at incorporating the two together.

**Comment by lyndsey-ferguson** [comment]
Let me know if you get stuck. I definitely want test_center to support this.

**Comment by schalkwijk** [comment]
@lyndsey-ferguson, didn't get stuck yet! [Here's](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/pull/258) my initial attempt at adding this to `test_center`.

**Comment by lyndsey-ferguson** [comment]
Nice, keep going!

File: lib/xctest_list.rb [hunk 1]
```diff
@@ -35,48 +35,50 @@ def self.binary_path(xctest_bundle_path)
...
-  def self.objc_tests(xctest_bundle_path)
+  def self.objc_tests(xctest_bundle_path, test_prefix: "test")
     tests = []
     objc_symbols_command_output_tempfile = Tempfile.new(File.basename(xctest_bundle_path) + "objc")
     system("nm -U '#{binary_path(xctest_bundle_path)}' > '#{objc_symbols_command_output_tempfile.path}'")
     tests = []
     File.foreach(objc_symbols_command_output_tempfile.path) do |line|
-      if / t -\[(?<testclass>\w+) (?<testmethod>test\w*)\]/ =~ line
-        tests << "#{testclass}/#{testmethod}"
+      regex = / t -\[(?<testclass>\w+) (?<testmethod>#{test_prefix}\w*)\]/
```
Comment (lyndsey-ferguson) [inline lib/xctest_list.rb:44]: So, the local variable assignment to `testmethod` will not work if there is variable interpolation in the regular expression: https://sadique.io/blog/2013/10/01/ruby-named-capture-groups-and-local-variables/

File: lib/xctest_list.rb [hunk 2]
```diff
@@ -35,48 +35,50 @@ def self.binary_path(xctest_bundle_path)
...
     swift_symbols_command_output_tempfile = Tempfile.new(File.basename(xctest_bundle_path) + "swift")
-    system("#{swift_symbol_dump_command(binary_path(xctest_bundle_path))}  > '#{swift_symbols_command_output_tempfile.path}'")
+    system("#{swift_symbol_dump_command(binary_path(xctest_bundle_path), test_prefix: test_prefix)}  > '#{swift_symbols_command_output_tempfile.path}'")
     tests = []
     File.foreach(swift_symbols_command_output_tempfile.path) do |line|
-      if /.*-\[.*\]/ !~ line && /\w+\.(?<testclass>[^\.]+)\.(?<testmethod>test[^\(]*)\(/ =~ line
-        tests << "#{testclass}/#{testmethod}"
+      regex = /\w+\.(?<testclass>[^\.]+)\.(?<testmethod>#{test_prefix}[^\(]*)\(/
+      if /.*-\[.*\]/ !~ line && matches = regex.match(line)
+        tests << "#{matches['testclass']}/#{matches['testmethod']}"
```
Comment (lyndsey-ferguson) [inline lib/xctest_list.rb:74]:
```suggestion
        tests << "#{matches[:testclass]}/#{matches[:testmethod]}"
```

File: lib/xctest_list.rb [hunk 3]
```diff
@@ -35,48 +35,50 @@ def self.binary_path(xctest_bundle_path)
...
     tests = []
     objc_symbols_command_output_tempfile = Tempfile.new(File.basename(xctest_bundle_path) + "objc")
     system("nm -U '#{binary_path(xctest_bundle_path)}' > '#{objc_symbols_command_output_tempfile.path}'")
     tests = []
     File.foreach(objc_symbols_command_output_tempfile.path) do |line|
-      if / t -\[(?<testclass>\w+) (?<testmethod>test\w*)\]/ =~ line
-        tests << "#{testclass}/#{testmethod}"
+      regex = / t -\[(?<testclass>\w+) (?<testmethod>#{test_prefix}\w*)\]/
+      if matches = regex.match(line)
+        tests << "#{matches['testclass']}/#{matches['testmethod']}"
```
Comment (lyndsey-ferguson) [inline lib/xctest_list.rb:46]:
```suggestion
        tests << "#{matches[:testclass]}/#{matches[:testmethod]}"
```

File: spec/xctest_list_spec.rb [hunk 4]
```diff
@@ -122,6 +122,13 @@ def mock_foreach(filepath_pattern, file_fixture)
       expect(parsed_tests).to eq(["AtomicBoyUITests/testExample2", "AtomicBoyUITests/testExample", "SwiftAtomicBoyUITests/testExample"])
     end
 
+    context 'when given a special test prefix' do
+        it 'finds swift 4 tests' do
```
Comment (lyndsey-ferguson) [inline spec/xctest_list_spec.rb:126]:
```suggestion
      it 'finds swift 4 tests' do
```

File: spec/xctest_list_spec.rb [hunk 5]
```diff
@@ -122,6 +122,13 @@ def mock_foreach(filepath_pattern, file_fixture)
       expect(parsed_tests).to eq(["AtomicBoyUITests/testExample2", "AtomicBoyUITests/testExample", "SwiftAtomicBoyUITests/testExample"])
     end
 
+    context 'when given a special test prefix' do
```
Comment (lyndsey-ferguson) [inline spec/xctest_list_spec.rb:125]: Just to keep things consistent in the file.

```suggestion
    describe 'when given a special test prefix' do
```

File: lib/xctest_list.rb (hunk 1 above)
Comment (schalkwijk) [reply lib/xctest_list.rb:44]: Thank you *so* much - I was struggling to find the reason why this was happening, but all of my googling turned up nothing.

File: spec/xctest_list_spec.rb (hunk 5 above)
Comment (schalkwijk) [reply spec/xctest_list_spec.rb:125]: Makes sense to me.

**Comment by schalkwijk** [comment]
@lyndsey-ferguson, all feedback addressed!

**Comment by lyndsey-ferguson** [comment]
Fixed in [v1.2.0](https://github.com/lyndsey-ferguson/xctest_list/releases/tag/v1.2.0)

**Comment by schalkwijk** [comment]
You're so fast! I'll work on `test_center` now.

**Pull request merged**